### DISCUSSION TITLE
[New feature] CCT - Correlated Color Temperature

### DISCUSSION THREAD

**Discussion by heinemannj** [Evercoder/culori#135] [Ideas]
Correlated Color Temperature (CCT) (in K)
Returns a correlated color temperature.

Option A:
Based on Bruce Justin Lindbloom's implementation
http://www.brucelindbloom.com/Eqn_XYZ_to_T.html

Option B:
based on McCamy cubic approximation
https://en.wikipedia.org/wiki/Color_temperature

```
var xyz65_Sum = xyz65.x + xyz65.y + xyz65.z;

if (xyz65_Sum !== 0) {
	xyY65.x = xyz65.x / xyz65_Sum;
	xyY65.y = xyz65.y / xyz65_Sum;
	xyY65.Y = xyz65.y;
} else {
	xyY65.x = 0;
	xyY65.y = 0;
	xyY65.Y = 0;
}

var n = (xyY65.x - 0.3320) / (0.1858 - xyY65.y);
var cct = (449.0 * Math.pow(n, 3)) + (3525.0 * Math.pow(n, 2)) + (6823.3 * n) + 5520.33;
```

**Benefit of option B:**

Caculation of **xyY coordinates** which are actually not covered by culori

**Comment by danburzo** [comment]
Thanks for opening this issue, CCT is one of the features I've been considering. In general, I would prefer a concise solution, where available, that doesn't rely on large lookup tables. I'll have to research the topic a bit.

**Comment by heinemannj** [comment]
With the help of below 2 functions I'm able to calculate the CCT in very flexible way:
(working for all White references (eg. D50, D65, ...)

```
function cct({mode, x, y, z}) {

	let xyY = convertXYZToxyY({mode, x, y, z});

	var n = (xyY.x - 0.3320) / (0.1858 - xyY.y);
	var res = (449.0 * Math.pow(n, 3)) + (3525.0 * Math.pow(n, 2)) + (6823.3 * n) + 5520.33;
	return res;
}
```

```
function convertXYZToxyY({mode, x, y, z}) {
	var sum = x + y + z;

	if (sum !== 0) {
		Y = y;
		x = x / sum;
		y = y / sum;
	} else {
		Y = 0;
		x = 0;
		y = 0;
	}

	let res = {
		mode: mode.replace("xyz", "xyY"),
		x: x,
		y: y,
		Y: Y
	}
	return res;
}
```

**For CCT calculation:**
`SensorData[4]['MeasuredValue']=parseFloat(cct(xyz65)).toFixed(0);`

**For transformation into CIE-xyY coordinates:**
`let xyY =convertXYZToxyY(xyz65));`

**Comment by danburzo** [comment]
This should also be an interesting read on the topic: https://www.osapublishing.org/oe/fulltext.cfm?uri=oe-24-13-14066&id=344803

**Comment by heinemannj** [comment]
It's up to you to choose from available solutions.

McCamy cubic approximation is very easy to implement - but for sure not the best solution in terms of accuracy

**Comment by heinemannj** [comment]
> Option A:
> Based on Bruce Justin Lindbloom's implementation
> http://www.brucelindbloom.com/Eqn_XYZ_to_T.html

Bruce Justin Lindbloom has created Javascript code based on Robertson method:
http://www.brucelindbloom.com/javascript/ColorConv.js

The deviation of the Robertson method (+/-12) compared with McCamy (+/- 284) is much more better.

Actually the most accurate solution seems to be the Y. Ohno's CCT (combined solution) calculation methods (+/-1)
http://www.bipublication.com/files/IJABR-V7I2-2016-29Zheleznikova.pdf

Phython script available on Colour's Git:
https://github.com/colour-science/colour/blob/develop/colour/temperature/ohno2013.py

**Comment by Artoria2e5** [comment]
The Ohno function requires at least a little bit of a lookup table because it requires working from Planck spectrum -- you'd at least need XYZ data points.

Robertson sounds more reasonable for the goal of not needing any tables. It's got some good simplicity and very nice use of the considerably more uniform mirad scale. There's even an [improved Robertson fit from 2022](https://www.tandfonline.com/doi/abs/10.1080/15502724.2023.2166060?journalCode=ulks20), going down to +/- 0.1 K from 1,500 K to 40,000 K; the number of data points is expanded from 31 to 372. [doug-baxter-modifications-of-the-robertson-method-for.pdf](https://github.com/Evercoder/culori/files/12403511/doug-baxter-modifications-of-the-robertson-method-for.pdf) At 188 points it's already better than Ohno, with max error of 0.4 K.


The above is for doing the xyY -> CCT calculation. For the inverse, which is what you need for guessing what a CCT looks like, you have two choices:

* Proper Planck. That would make Ohno 0-cost in terms of data.
* Kang et al. (2002), https://github.com/colour-science/colour/blob/develop/colour/temperature/kang2002.py, paper available at https://www.kci.go.kr/kciportal/ci/sereArticleSearch/ciSereArtiView.kci?sereArticleSearchBean.artiId=ART000987865 (use the button on the right with the PDF icon) &ndash; this actually also covers xyY -> CCT, it just doesn't give any ready-to-use fit.

**Comment by dlamon1** [comment]
Is there anything an uneducated layman, such as myself, do to help implement this feature into the library?
